apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java

```diff
@@ -0,0 +1,59 @@
+package gov.cms.bfd.pipeline.rda.grpc.server;
+
+import java.util.NoSuchElementException;
+import java.util.function.Predicate;
+
+/**
+ * Wrapper around a {@link MessageSource} to allow the values to be filtered using some condition.
+ * Generally the filter will be a minimum sequence number.
+ *
+ * @param <T> type of objects being returned by the source.
+ */
+public class FilteredMessageSource<T> implements MessageSource<T> {
+  private final MessageSource<T> source;
+  private final Predicate<T> filter;
+  private T nextValue;
+
+  /**
+   * Wrap the specified source and filter its objects using the supplied predicate. The predicate
+   * must return true for objects that are valid and false for objects that are invalid.
+   *
+   * @param source real source of objects
+   * @param filter function returning true for objects to keep or false for objects to skip
+   */
+  public FilteredMessageSource(MessageSource<T> source, Predicate<T> filter) {
+    this.source = source;
+    this.filter = filter;
```

Comment:
Should we advance the pointer here so that `next()` can be called?

Reply:
No it's part of the design that the constructor doesn't do any IO.  That happens in `hasNext()`.